pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+	memoryLimit uint64
+	// CPU limit in cores
+	cpuLimit float64
+	// Last CPU time counter
+	lastCPUTime float64
+	// The time of the last update
+	lastUpdate     time.Time
+	movingAvg      *math.EwmaRate
+	LimitingReason atomic.String
+}
+
+// NewUtilizationBasedLimiter returns a UtilizationBasedLimiter configured with cpuLimit and memoryLimit.
+func NewUtilizationBasedLimiter(cpuLimit float64, memoryLimit uint64, logger log.Logger) *UtilizationBasedLimiter {
+	// Calculate alpha for a minute long window
+	// https://github.com/VividCortex/ewma#choosing-alpha
+	alpha := 2 / (60/resourceUtilizationUpdateInterval.Seconds() + 1)
+	l := &UtilizationBasedLimiter{
+		logger:      logger,
+		cpuLimit:    cpuLimit,
+		memoryLimit: memoryLimit,
+		// Use a minute long window, each sample being a second apart
+		movingAvg: math.NewEWMARate(alpha, resourceUtilizationUpdateInterval),
+	}
+	l.Service = services.NewTimerService(resourceUtilizationUpdateInterval, l.init, l.update, nil)
+	return l
+}
+
+func (l *UtilizationBasedLimiter) init(_ context.Context) error {
+	s, ok := newCgroupV2Scanner()
+	if ok {
+		l.utilizationScanner = s
+		return nil
+	}
+	s, ok = newCgroupV1Scanner()
+	if ok {
+		l.utilizationScanner = s
+		return nil
+	}
+	s, ok = newProcfsScanner()
+	if ok {
+		l.utilizationScanner = s
+		return nil
+	}
+
+	return fmt.Errorf("unable to detect CPU/memory utilization, unsupported platform")
+}
+
+func (l *UtilizationBasedLimiter) update(ctx context.Context) error {
+	cpuTime, memUtil, err := l.utilizationScanner.Scan()
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "failed to get CPU and memory stats", "method",
+			l.utilizationScanner.String(), "err", err.Error())
+		// Disable any limiting, since we can't tell resource utilization
+		l.LimitingReason.Store("")
+		return ctx.Err()
+	}
+
+	now := time.Now().UTC()
+	lastUpdate := l.lastUpdate
+	l.lastUpdate = now
+
+	lastCPUTime := l.lastCPUTime
+	l.lastCPUTime = cpuTime
+
+	if lastUpdate.IsZero() {
+		return ctx.Err()
+	}
+
+	cpuUtil := (cpuTime - lastCPUTime) / now.Sub(lastUpdate).Seconds()
+	l.movingAvg.Add(int64(cpuUtil * 100))
+	l.movingAvg.Tick()
+	cpuA := float64(l.movingAvg.Rate()) / 100
+
+	level.Debug(l.logger).Log("msg", "process resource utilization", "method", l.utilizationScanner.String(),
+		"memory_utilization", memUtil, "smoothed_cpu_utilization", cpuA, "raw_cpu_utilization", cpuUtil)
+
+	memPercent := 100 * (float64(memUtil) / float64(l.memoryLimit))
+	cpuPercent := 100 * (cpuA / l.cpuLimit)
+
+	var reason string
+	if memPercent >= 100 {
+		reason = "memory"
+	} else if cpuPercent >= 100 {
+		reason = "cpu"
+	}
+
+	enable := reason != ""
+	prevEnable := l.LimitingReason.Load() != ""
+	if enable == prevEnable {
+		// No change
+		return ctx.Err()
+	}
+
+	if enable {
+		level.Info(l.logger).Log("msg", "enabling resource utilization based limiting",
+			"reason", reason, "memory_limit", l.memoryLimit,
+			"memory_percentage_of_limit", memPercent, "cpu_limit", l.cpuLimit,
```

Comment:
[nit] I would remove the "percentage" and just show the "cpu_utilization" and "memory_utilization", like you do for the limit. Than the reader can do math itself, if needed. This allows you to remove `memPercent` and `cpuPercent` and simply the check if limit has been reached to:

```
	if memUtil > l.memoryLimit {
		reason = "memory"
	} else if cpuA > l.cpuLimit {
		reason = "cpu"
	}
```

_Comment applies below too._

Reply:
Done.